File: Algorithm/040. Combination Sum II.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-04-12 10:52:14
# @Last modified by:   WuLC
# @Last Modified time: 2016-04-13 09:29:04
# @Email: liangchaowu5@gmail.com

# DP
# candidates中的元素只能用一次，但是candidates中可能有重复的元素
class Solution(object):
    def combinationSum2(self, candidates, target):
    	can_dict = {}
    	for candidate in candidates:
    		if candidate not in can_dict: 
    		    can_dict[candidate] = 0
    		can_dict[candidate] += 1

        dp = {}
        dp[0] = []
        for i in xrange(1,target+1):
            dp[i] = []
            if i in candidates:
                dp[i].append([i])
            for j in xrange(1,i/2+1):
                if len(dp[j])==0 or len(dp[i-j])==0:
                    continue
                for m in dp[j]:
                    for n in dp[i-j]:
                        tmp_dict = {}
                        flag = 0               
                        tmp = m+n
                        for t in tmp:
                            if t not in tmp_dict: 
                                tmp_dict[t] = 0
                            tmp_dict[t] += 1
                       
                        for te in tmp_dict:
                            if tmp_dict[te] > can_dict[te]: 
	                            flag = 1
	                            break
                        if flag == 0:
                            tmp.sort()                         
                            if tmp not in dp[i]:
                                dp[i].append(tmp)
        return dp[target]

# bracktracking
class Solution(object):
    def combinationSum2(self, candidates, target):
        candidates.sort()
        result = []
        self.dfs(candidates,target,0 ,[],result)
        return result


    def dfs(self, nums, tar, index, tmp, res):
        if tar == 0 :
            if tmp not in res:
                res.append(tmp)
        else:
            for i in range(index,len(nums)):
                if tar-nums[i]<0:
                    return
                self.dfs(nums, tar-nums[i], i+1, tmp+[nums[i]], res)

File: Algorithm/Java/040. Combination Sum II.java
/**
* Author: WuLC
* Date:   2016-12-08 18:54:44
* Last modified by:   WuLC
* Last Modified time: 2016-12-08 18:55:40
* Email: liangchaowu5@gmail.com
*/

// backtracking 
public class Solution 
{
    private ArrayList<List<Integer>> result;  
    public List<List<Integer>> combinationSum2(int[] candidates, int target) 
    {
        result = new ArrayList<List<Integer>>();
        List<Integer> tmp = new ArrayList<Integer>();
        Arrays.sort(candidates);
        dfs(candidates, tmp, 0, 0, target);
        return result;
    }
    
    public void dfs(int[] candidates, List<Integer> tmp, int idx, int count, int target)
    {
        if (count == target)
        {
            result.add(new ArrayList(tmp));
            return ;
        }
        
        for (int i=idx; i < candidates.length; i++)
        {
            if (i > idx && candidates[i] == candidates[i-1]) continue; // deal with duplicate numbers
            if (count+candidates[i] <= target)
            {
                tmp.add(candidates[i]);
                dfs(candidates, tmp, i+1, count+candidates[i], target);
                tmp.remove(tmp.size()-1);
            }
            else return;
        }
    }
}